src/blocks/battery/apc_ups.rs

```diff
@@ -0,0 +1,196 @@
+use std::str::FromStr;
+use tokio::net::TcpStream;
+use tokio::time::Interval;
+
+use super::{BatteryDevice, BatteryInfo, BatteryStatus, DeviceName};
+use crate::blocks::prelude::*;
+
+#[derive(Debug, Default)]
+struct PropertyMap(HashMap<String, String>);
+
+make_log_macro!(debug, "battery[apc_ups]");
+
+impl PropertyMap {
+    fn insert(&mut self, k: String, v: String) -> Option<String> {
+        self.0.insert(k, v)
+    }
+
+    fn get(&self, k: &str) -> Option<&String> {
+        self.0.get(k)
+    }
+
+    fn get_property<T: FromStr + Send + Sync>(
+        &self,
+        property_name: &str,
+        required_unit: &str,
+    ) -> Result<T> {
+        let stat = self
+            .get(property_name)
+            .or_error(|| format!("{} not in apc ups data", property_name))?;
+
+        let (value, unit) = stat
+            .split_once(' ')
+            .or_error(|| format!("could not split {}", property_name))?;
+        if unit == required_unit {
+            value
+                .parse::<T>()
+                .map_err(|_| Error::new("Could not parse data"))
+        } else {
+            return Err(Error::new(format!(
+                "Expected unit for {} are {}, but got {}",
+                property_name, required_unit, unit
+            )));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ApcConnection(TcpStream);
+
+impl ApcConnection {
+    async fn connect(addr: String) -> Result<Self> {
+        Ok(Self(
+            TcpStream::connect(addr)
+                .await
+                .error("Failed to connect to socket")?,
+        ))
+    }
+
+    async fn write(&mut self, msg: &[u8]) -> Result<()> {
+        let msg_len = u16::try_from(msg.len())
+            .error("msg is too long, it must be less than 2^16 characters long")?;
+
+        self.0
+            .write_u16(msg_len)
+            .await
+            .error("Could not write message length to socket")?;
+        self.0
+            .write_all(msg)
+            .await
+            .error("Could not write message to socket")?;
+        Ok(())
+    }
+
+    async fn read_response(&mut self) -> Result<String> {
+        let mut buf = String::new();
+        loop {
+            let read_size = self
+                .0
+                .read_u16()
+                .await
+                .error("Could not read response length from socket")?
+                .into();
+            if read_size == 0 {
+                break;
+            }
+            let mut read_buf = vec![0_u8; read_size];
+            self.0
+                .read_exact(&mut read_buf)
+                .await
+                .error("Could not read from socket")?;
+            buf.extend(String::from_utf8(read_buf));
```

Comment:
Probably the last comment. Here `String::from_utf8()` returns `Result`, which happens to implement `IntoIterator`. So this code is equivalent to

```rust
if let Ok(s) = String::from_utf8(read_buf) {
    buf.push_str(&s);
}
```

You probably wanted something like

```rust
buf.push_str(std::str::from_utf8(&read_buf).error(...)?);
```

Reply:
fixed in my commit